elixir/apps/domain/lib/domain/policies.ex

```diff
@@ -196,21 +196,61 @@ defmodule Domain.Policies do
     end)
   end
 
-  def ensure_client_conforms_policy_conditions(%Clients.Client{} = client, %Policy{} = policy) do
-    case Condition.Evaluator.ensure_conforms(policy.conditions, client) do
-      {:ok, expires_at} ->
-        {:ok, expires_at}
+  @infinity ~U[9999-12-31 23:59:59.999999Z]
 
-      {:error, violated_properties} ->
-        {:error, {:forbidden, violated_properties: violated_properties}}
+  def longest_conforming_policy_for_client(policies, client, token_expires_at) do
+    policies
+    |> Enum.reduce(%{failed: [], succeeded: []}, fn policy, acc ->
+      case ensure_client_conforms_policy_conditions(client, policy) do
+        {:ok, expires_at} ->
+          %{acc | succeeded: [{expires_at, policy} | acc.succeeded]}
+
+        {:error, {:forbidden, violated_properties: violated_properties}} ->
+          %{acc | failed: acc.failed ++ violated_properties}
+      end
+    end)
+    |> case do
+      %{succeeded: [], failed: failed} ->
+        {:error, {:forbidden, violated_properties: Enum.uniq(failed)}}
+
+      %{succeeded: succeeded} ->
+        {expires_at, policy} =
+          succeeded
+          |> Enum.max_by(fn {expires_at, _policy} -> expires_at || @infinity end)
+
+        {:ok, min_expires_at(expires_at, token_expires_at), policy}
+    end
+  end
+
+  # All client tokens have *some* expiration
+  def min_expires_at(nil, nil),
+    do: raise("Both policy_expires_at and token_expires_at cannot be nil")
+
+  def min_expires_at(nil, token_expires_at), do: token_expires_at
+
+  def min_expires_at(%DateTime{} = policy_expires_at, %DateTime{} = token_expires_at) do
+    if DateTime.compare(policy_expires_at, token_expires_at) == :lt do
+      policy_expires_at
+    else
+      token_expires_at
     end
   end
 
-  def ensure_has_access_to(%Auth.Subject{} = subject, %Policy{} = policy) do
+  defp ensure_has_access_to(%Auth.Subject{} = subject, %Policy{} = policy) do
```

Comment:
Just curious why this is being made private?

Reply:
It's not called from outside the module.

Reply:
We probably have many more like this. This creates a problem because static analysis can't determine if this function is no longer used.

It will only alert us to unused private functions.

Reply:
Gotcha.  I use some (possibly all) of these types of functions in the `hard-delete` branch I'm working on, which is why I was asking.  I can always make it public again if needed.

Reply:
I see you guys are discovering type systems and static analysis 😁😋